lib/hierarchical.h

```diff
@@ -465,6 +477,117 @@ class HierarchicalRadiosity {
         }
     }
 
+    /**
+     * Refine all links in all nodes.
+     * @return true if links have been refined.
+     */
+    bool refine_links() {
+        bool done = true;
+        for (auto& p : nodes_) {
+            if(!refine_links(p)) {
+                done = false;
+            }
+        }
+        return done;
+    }
+
+    /**
+     * Refine all links in node p.
+     * @return true if a link has been refined.
+     */
+    bool refine_links(Quadnode& p) {
+        bool done = true;
+
+        // We create a copy of gathering links because refine_link might remove
+        // elements.
+        // TODO: Avoid the copy
+        std::vector<Linknode> links = p.gathering_from;
+        for (auto& link: links) {
+            if (!refine_link(p, link)) {
```

Comment:
Du koenntest hier ueber Indizes von links iterieren. In refine_link würde ich den link nicht löschen. Wenn ich das richtig verstehe, falls refine_link neue Links erstellt, dann erstellt es die auf einer Ebene tiefer, insbesondere in einer anderen Liste. D.h. hier, falls du einen Link löschen willst, kannst du den mit dem letzten Link swappen und den Vector um eins verkleinern.